### PR TITLE
(PUP-8635) Add specification of the Deferred data type

### PR DESCRIPTION
**Pull request by hlindberg**

### CODE REVIEW

File: language/types_values_variables.md [hunk 1]
```diff
@@ -2185,6 +2188,52 @@ There is no automatic unwrapping of sensitive values. As a consequence it is not
...
+
+Creates a new instance of the `Deferred` type. It accepts the name of the function to later call, and an array of
+arguments to use when making the call. The arguments are evaluated at the time the deferred value is created, but
+any deferred values among the arguments will be resolved as part of the resolution process (when actually making the call).
+
+    $x = Deferred('some_function', ['argument1', 42])
+
+Will, when the deferred is resolved result in the call:
+
+    some_function('argument1', 42)
```
Comment (ripienaar) [inline language/types_values_variables.md]: and presumably these would be ruby functions? or pp functions too? Or something else entirely (would they have scope and facts?)

Comment (hlindberg) [reply language/types_values_variables.md]: In first release only ruby functions. They are called in a global scope. `$facts` is set with the facts when deferred is resolved on the agent side.
(If resolved while compiling - if you call the deferred, it has everything the compiler has).

**Comment by hlindberg** [comment]
The specification should include a "since puppet 6.0.0" for Deferred.

File: language/types_values_variables.md (hunk 1 above)
Comment (ripienaar) [reply language/types_values_variables.md]: nice, ok thanks

Comment (hlindberg) [reply language/types_values_variables.md]: PR updated with more details

**Pull request merged**